stylix/testbed/autoload.nix

```diff
@@ -0,0 +1,118 @@
+{
+  pkgs,
+  lib,
+  testbedFieldSeparator ? ":",
+}:
+let
+  isEnabled = pkgs.callPackage ./is-enabled.nix { };
+
+  # Describes all testbed modules loaded from the modules directory.
+  # A list of attrsets, each containing: { module, path, name }
+  testbeds = lib.pipe ../../modules [
+    builtins.readDir
+    builtins.attrNames
+    (builtins.concatMap (
+      module:
+      let
+        testbeds = ../../modules/${module}/testbeds;
+        files = lib.optionalAttrs (builtins.pathExists testbeds) (
+          builtins.readDir testbeds
+        );
+      in
+      lib.mapAttrsToList (
+        testbed: type:
+        let
+          path = testbeds + "/${testbed}";
+          pathStr = toString path;
+        in
+        if type != "regular" then
+          throw "${pathStr} must be regular: ${type}"
+
+        else if !lib.hasSuffix ".nix" testbed then
+          throw "testbed must be a Nix file: ${pathStr}"
+
+        else if testbed == ".nix" then
+          throw "testbed must have a name: ${pathStr}"
+
+        else
+          {
+            inherit module path;
+            name = lib.removeSuffix ".nix" testbed;
+          }
+      ) files
+    ))
+  ];
+
+  # Import all the testbed themes
+  themes = lib.pipe ./themes [
+    builtins.readDir
+    (lib.filterAttrs (name: _: lib.strings.hasSuffix ".nix" name))
+    (builtins.mapAttrs (
+      name: type:
+      lib.throwIfNot (type == "regular")
+        "Unexpected filetype in testbed themes: ${toString ./themes/${name}} is a ${type}."
+        ./themes/${name}
+    ))
+    (lib.mapAttrs' (name: lib.nameValuePair (lib.strings.removeSuffix ".nix" name)))
+  ];
+
+  # Construct a NixOS module for the testbed+theme combination.
+  #
+  # If the testbed module is enabled, returns an attrset containing it:
+  # { «name» = «modele»; }
+  #
+  # Returns an empty attrset if the testbed module is not enabled.
+  # { }
+  makeTestbedModule =
+    testbed: themeName: themeModule:
+    let
+      joinFields = builtins.concatStringsSep testbedFieldSeparator;
+
+      fields =
+        map
+          (
+            field:
+            lib.throwIf (lib.hasInfix testbedFieldSeparator field)
+              "testbed field must not contain the '${testbedFieldSeparator}' testbed field separator: ${field}"
+              field
+          )
+          [
+            "testbed"
+            testbed.name
+            themeName
+          ];
+
+      name = joinFields fields;
+    in
+    lib.optionalAttrs (isEnabled testbed.path) {
+      ${name} = {
+        # Unique key for de-duplication
+        key = joinFields [
+          (toString ./.)
+          name
+        ];
```

Comment:
Sorry for missing the obvious, but why are we adding this `key` attribute? And what exactly is being de-duplicated? Also, would the `toString ./.` prefix not be the same for every `key`, making it redundant regarding uniqueness? In that case, `key` is as unique as the `${name}` attribute name that is already required to be unique due to how Nix attribute sets work.